core/validators/signatures.go

```diff
@@ -99,184 +127,252 @@ func (s *ERC20Signatures) EmitPromotionsSignatures(
-		if val.Status == ValidatorStatusTendermint {
-			allValidators = append(allValidators, NodeIDAddress{k, val.EthAddress})
-		}
 		if _, ok := previousState[k]; !ok {
 			// this is a new validator which didn't exist before
 			if val.Status == ValidatorStatusTendermint {
-				toAdd = append(toAdd, NodeIDAddress{k, val.EthAddress})
+				toAdd = append(toAdd, NodeIDAddress{k, val.EthAddress, val.SubmitterAddress})
 			}
 		}
 	}
 
 	s.SetNonce(currentTime)
-	s.EmitNewValidatorsSignatures(ctx, toAdd, currentTime, epochSeq)
-	s.EmitRemoveValidatorsSignatures(ctx, toRemove, allValidators, currentTime, epochSeq)
+	s.PrepareValidatorSignatures(ctx, toAdd, epochSeq, true)
+	s.PrepareValidatorSignatures(ctx, toRemove, epochSeq, false)
+
+	// check if the node being added has supplied a submitterAddress because if it has we can automatically emit
+	// signatures for the node to add itself
+	for _, v := range toAdd {
+		if v.SubmitterAddress != "" {
+			s.log.Debug("sending automatic add signatures", logging.String("submitter", v.SubmitterAddress), logging.String("nodeID", v.NodeID))
+			s.EmitValidatorAddedSignatures(ctx, v.SubmitterAddress, v.NodeID, currentTime)
+		}
+	}
+
+	// for each node being removed check if the Tendermint nodes have a submitter address because if they do
+	// we can automatically emit remove signatures for them
+	for _, r := range toRemove {
+		for _, v := range newState {
+			if v.SubmitterAddress != "" && v.Status == ValidatorStatusTendermint {
+				s.log.Debug("sending automatic remove signatures", logging.String("submitter", v.SubmitterAddress), logging.String("nodeID", r.NodeID))
+				s.EmitValidatorRemovedSignatures(ctx, v.SubmitterAddress, r.NodeID, currentTime)
+			}
+		}
+	}
 }
 
 func (s *ERC20Signatures) SetNonce(t time.Time) {
 	s.lastNonce = num.NewUint(uint64(t.Unix()) + 1)
 }
 
-func (s *ERC20Signatures) EmitNewValidatorsSignatures(
-	ctx context.Context,
-	validators []NodeIDAddress,
-	currentTime time.Time,
-	epochSeq uint64,
-) {
+// PrepareValidatorSignatures make nonces and store the data needed to generate signatures to add/remove from the multisig control contract.
+func (s *ERC20Signatures) PrepareValidatorSignatures(ctx context.Context, validators []NodeIDAddress, epochSeq uint64, added bool) {
 	sort.Slice(validators, func(i, j int) bool {
 		return validators[i].EthAddress < validators[j].EthAddress
 	})
-	evts := []events.Event{}
 
 	for _, signer := range validators {
+		d := &signatureData{
+			NodeID:     signer.NodeID,
+			EthAddress: signer.EthAddress,
+			Nonce:      s.lastNonce.Clone(),
+			EpochSeq:   epochSeq,
+			Added:      added,
+		}
+		s.lastNonce.AddUint64(s.lastNonce, 1)
+
+		// we're ok to override whatever is in here since we can't need to both add and remove an eth-address at the same time
+		// so we'll replace it either with the correct action, or with the same action but with a later epoch which is fine
+		// because you can still get signatures to do what you need
```

Comment:
Is that true? E.g in the case of an ethereum key rotation?

Reply:
Its true because the map is from `ethereum-address -> the sig data` (not `nodeID -> sig data`) so when you rotate your key the address would change so you'd have one element for the old-key with remove, and one for the new-key with add.

Reply:
yup I realised that down the line! thanks.